### ISSUE TITLE
No sign of current BBC Radio shows in 10.12.0

### ISSUE THREAD

**Issue by Necktilt** [Ascoware/get-iplayer-automator#144]
Just updated to 10.12.0 and searching for a few of last week's radio shows drew a blank. For example, Keywords In Our Time, R4, Mon 19 March and next four days. No changes in my prefs. Worked OK when I did a Use Current Webpage. Anything else changed to create this?

**Comment by PhoolR** [comment]
Try a force cache update first? Then it might work. It has all been working for me since the update without having to force.

**Comment by sleaux-meaux** [comment]
The programme name is Keywords FOR Our Time.

**Comment by Necktilt** [comment]
Phil, forced cache update, no change.

S-m, sorry, my mistype there, but actually I searched for just "Keywords" without result. 

Also, for example, Desert Island Discs with David Byrne.

**Comment by Necktilt** [comment]
When I force a cache update, I see this in the log:

INFO: Indexing radio programmes (concurrent)
WARNING: Failed to load LWP SSL options
WARNING: SSL hostname verification disabled for cache update

WARNING: Failed to load LWP SSL options
WARNING: SSL hostname verification disabled for cache update

INFO: Added 0 radio programmes to cache
AppController: Index Updated.

**Comment by sleaux-meaux** [comment]
Those warnings doesn't necessarily mean cache updates will fail, but they indicate something is wrong with the perl on your system. Turn on verbose mode and there should be some more information with those warnings.

**Comment by sleaux-meaux** [comment]
Sorry - scratch that. I just tried on my machine and the additional information doesn't appear in the GiA log window. GiA doesn't pass the verbose mode flag to get_iplayer for cache updates. Outside of GiA I can see that, as I suspected, the built-in perl on macOS is missing a module necessary for verifying HTTPS connections. Still, that shouldn't prevent get_iplayer from updating the cache, so back to square one.

**Comment by PhoolR** [comment]
I've just done an update (normal) and it has added all the radio shows I expected it to add. (last one was from this morning).

**Comment by Necktilt** [comment]
Just opened GiA for another look, log says ...

<< Get iPlayer Automator 1.12.0 Initialized.
AppController: Updating Programme Index Feeds...

INFO: Indexing radio programmes (concurrent)
WARNING: Failed to load LWP SSL options
WARNING: SSL hostname verification disabled for cache update

WARNING: Failed to load LWP SSL options
WARNING: SSL hostname verification disabled for cache update

INFO: Added 136 radio programmes to cache >>

... So it's indexing new radio progs. But I'm still not seeing stuff in search from last week's radio. (I only use GiA for radio.) But ... I do get programmes from this week (in other words today). For example Start The Week, Love and Loss, or 15 Minute Drama, Judas, Episode 1. So it looks like I just seem to be missing last week's. Weird, no?

**Comment by sleaux-meaux** [comment]
It depends on when you updated your cache over the past week, and with which version of GiA. You can back up and delete your radio cache file and launch GiA to regenerate the cache back through the 19th. However, you will lose cache entries for programmes before the 19th, so perhaps not worth the bother if you already have the programmes you want. If you're feeling adventurous, you can rebuild your BBC caches in their entirely with this (assuming GiA installed in /Applications):

    PERL5LIB="/Applications/Get iPlayer Automator.app/Contents/Resources/perl5" "/Applications/Get iPlayer Automator.app/Contents/Resources/get_iplayer" --profile-dir="~/Library/Application Support/Get iPlayer Automator" --rebuild-cache --type=tv,radio

**Comment by Necktilt** [comment]
Thanks s-m. Think I'll play it safe, as I do have what I want for now, and it looks like going ahead will be fine. Though I say that with fingers crossed. Thanks again to you and Phil for the help: appreciated.

**Comment by sleaux-meaux** [comment]
Actually, a better solution would be to reset the timestamp on your radio cache file to a time from last week, which will force get_iplayer to refresh back through the 19th the next time it runs, without losing earlier entries:

    touch -t 201803191200 "$HOME/Library/Application Support/Get iPlayer Automator/radio.cache"

**Comment by Vangelis66** [comment]
> WARNING: Failed to load LWP SSL options
> WARNING: SSL hostname verification disabled for cache update

> but they indicate something is wrong with the perl on your system.

> the built-in perl on macOS is missing a module necessary for verifying HTTPS connections.

@sleaux-meaux 

All the above is the result of commit https://github.com/get-iplayer/get_iplayer/commit/df0d32b1fc946defe80ea356f44565df56a92dd5 that the get_iplayer dev(s) pushed to v**3.13**... 

> It depends on when you updated your cache over the past week, and with which version of GiA. 
> You can back up and delete your radio cache file and launch GiA to regenerate the cache back 
> through the 19th. However, you will lose cache entries for programmes before the 19th, 
> so perhaps not worth the bother.

> Actually, a better solution would be to reset the timestamp on your radio cache file to a time from last 
> week, which will force get_iplayer to refresh back through the 19th the next time it runs this week, 
> without losing earlier entries:

The standalone get_iplayer perl script has the potential to refresh the (tv|radio).cache in a retrograde fashion, using the `--refresh-limit <days>` switch; e.g. 

`get_iplayer --type=tv,radio -f --refresh-limit=15 --force` 

will (re-)populate the caches with all content from the last two weeks (without removing older existing entries); also, 

`get_iplayer --type=tv,radio --cache-rebuild`

will rebuild a full 30d-worth cache. 
Do you think this info could be of use/applicable to GiA users, too? 

Regards

**Comment by skovatch** [comment]
I am inclined to put in a 'Rebuild BBC cache' with a number-of-days option, though it seems like something you would only need to do in extreme circumstances.

The current ITV cache will fetch everything available on ITV Hub, and I don't see a reason to change that since it's reasonably fast to do it all the time.

**Comment by skovatch** [comment]
The warnings, IMO, are harmless, though I appreciate why they were added. If you are going to make secure connections it makes sense to verify you are connecting to the server you think you are connecting to. I added the necessary perl library that makes the message go away.

**Comment by skovatch** [comment]
The warnings are now fixed in v1.12.1. I haven't done anything about cache rebuilds, though. I do see the cache being updated so I'm closing this as fixed.

**Comment by sleaux-meaux** [comment]
> All the above is the result of commit get-iplayer/get_iplayer@df0d32b 

That code merely serves to make the module used to perform cache updates (Mojo::UserAgent) behave the same as the module used to perform other downloads (LWP::UserAgent) with respect to SSL hostname verfication (verification on by default). You need some CA certificates to do that, but for quite some time Apple has shipped a broken version of LWP that doesn't have access to such certificates (thus the warnings). This is coming to the fore with get_iplayer as the BBC begins redirecting URLs to HTTPS equivalents. There are different ways to deal with the situation, with the most expedient being the installation of an additional module (Mozilla::CA). GiA 1.12.1 embeds this module, which is slightly sub-optimal because you're accepting certificates from an intermediary, but embedding is the only way to retain drag-and-drop installation. You can verify the supplied certificates file against the source, but the Mozilla::CA module is itself an intermediary of sorts. The embedded module is also incomplete because it omits the script you could use to create a certificates file yourself, but that script isn't necessarily completely secure. So it's turtles all the way down. I realise that no GiA users are likely to care about any of this, but fyi.

> get_iplayer --type=tv,radio --cache-rebuild

I included a GiA equivalent in an earlier post, but removed it because it was overkill for the problem at hand. For the curious:

    PERL5LIB="/Applications/Get iPlayer Automator.app/Contents/Resources/perl5" "/Applications/Get iPlayer Automator.app/Contents/Resources/get_iplayer" --profile-dir="$HOME/Library/Application Support/Get iPlayer Automator" --rebuild-cache --type=tv,radio
    
If you are not able to update your cache on a weekly basis and you want to recover older programmes missing from search results, use the above. Unless your internet connection is very slow, you don't need to figure out how many days to go back. Just rebuild everything. If you really want to figure out how old your cache files are, replace `--rebuild-cache` with `--show-cache-age` in the command above. Ages are given in seconds, so divide by 86400, toss the remainder, and add 1 to get a value to use with `--refresh-limit`.

**Comment by Necktilt** [comment]
S-m, I do generally update the cache weekly, so I guess I'm OK for now without fiddling. 

In parallel, sort of, I did have some trouble with https/BBC stuff and Logitech Media Server, but updating to the latest version of that sorted it, so presumably they've dealt with that too.

**Comment by skovatch** [comment]
This is good advice... I need to get this site's Wiki organized and up and running so we can capture things like this.

**Comment by Necktilt** [comment]
Sleaux-meaux & Scott

I upgraded to 1.12.2 in the meantime, but thought you ought to know that today when I went to get some radio shows from last week, Sat 24th onward, I had the same problem that I reported when I started the thread: in other words, nothing coming up in the search for that week. 

Plenty before and after, but nothing from that week. Forced cache update, still nothing. Tried s-m’s PERL5LIB trick, and that did it. 

So I’m still curious as to what has changed in the app that’s made this blip in performance. Would be nice to have it back doing what it always did before!

All the best,

Tony
